rpm/0005-Support-OneDrive-oauth2-flow.patch

```diff
@@ -0,0 +1,146 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Chris Adams <chris.adams@jollamobile.com>
+Date: Thu, 2 Apr 2020 14:30:48 +1000
+Subject: [PATCH] Support OneDrive oauth2 flow
+
+OneDrive supports a slightly non-standard USER_AGENT flow where
+a bearer code may be exchanged for a token.
+
+Also, support using a dynamically determined redirect uri for
+the authorization code request.
+---
+ src/oauth2plugin.cpp | 98 ++++++++++++++++++++++++++------------------
+ 1 file changed, 58 insertions(+), 40 deletions(-)
+
+diff --git a/src/oauth2plugin.cpp b/src/oauth2plugin.cpp
+index 3eb3fef..dc1dd7e 100644
+--- a/src/oauth2plugin.cpp
++++ b/src/oauth2plugin.cpp
+@@ -395,7 +395,11 @@ void OAuth2Plugin::userActionFinished(const SignOn::UiSessionData &data)
+ 
+     if (handleUiErrors(data)) return;
+ 
+-    TRACE() << data.UrlResponse();
++    const QString redirectUri = data.getProperty(QLatin1String("RedirectUri")).toString().isEmpty()
++                              ? d->m_oauth2Data.RedirectUri()
++                              : data.getProperty(QLatin1String("RedirectUri")).toString();
++
++    TRACE() << "OAuth2Plugin:userActionFinished: got:" << data.UrlResponse() << redirectUri;
+ 
+     // Checking if authorization server granted access
+     QUrl url = QUrl(data.UrlResponse());
+@@ -406,50 +410,63 @@ void OAuth2Plugin::userActionFinished(const SignOn::UiSessionData &data)
+     }
+ 
+     if (d->m_mechanism == USER_AGENT) {
+-        // Response should contain the access token
+-        OAuth2PluginTokenData respData;
+-        if (url.hasFragment()) {
+-            QString state;
+-            respData.setScope(d->m_oauth2Data.Scope());
+-            QUrlQuery fragment(url.fragment());
+-            QVariantMap extraFields;
+-            typedef QPair<QString, QString> StringPair;
+-            Q_FOREACH(const StringPair &pair, fragment.queryItems()) {
+-                if (pair.first == ACCESS_TOKEN) {
+-                    respData.setAccessToken(pair.second);
+-                } else if (pair.first == ID_TOKEN) {
+-                    respData.setIdToken(pair.second);
+-                } else if (pair.first == EXPIRES_IN) {
+-                    respData.setExpiresIn(pair.second.toInt());
+-                } else if (pair.first == REFRESH_TOKEN) {
+-                    respData.setRefreshToken(pair.second);
+-                } else if (pair.first == STATE) {
+-                    state = pair.second;
+-                } else if (pair.first == SCOPE) {
+-                    respData.setScope(pair.second.split(' ', QString::SkipEmptyParts));
++        // some providers (e.g. OneDrive) implement non-standard OAuth2 flows
++        // and return auth code requiring exchange for token.
```

Comment:
I'm just wondering if this should now add `state` into the url. Previously `state` was set by signonui. Or if possible we'd drop this patch but I don't know if OneDrive still needs this.

Reply:
Looking more carefully, I'm confident that this works like it should, assuming the old version used to work. Just the question if we could drop the patch remains.